Overriding Panache methods leads to duplicated method error

**Describe the bug**
I have a class implementng a PanacheRepositoryBase<T, P> 

```
@ApplicationScoped
public class JpaOwnerRepository implements PanacheRepositoryBase<Owner, Integer> {

    @Inject
    EntityManager em;

    @Override
    public Owner findById(Integer id) {
        // using 'join fetch' because a single query should load both owners and pets
        // using 'left join fetch' because it might happen that an owner does not have
        // pets yet
        Query query = this.em
                .createQuery("SELECT owner FROM Owner owner left join fetch owner.pets WHERE owner.id =:id");
        query.setParameter("id", id);
        return (Owner) query.getSingleResult();
    }

    @Audited
    public void save(Owner owner) {
        if (owner.getId() == null) {
            persist(owner);
        } else {
            this.em.merge(owner);
        }

    }
}
```

Error received 
```
org.junit.jupiter.api.extension.TestInstantiationException: TestInstanceFactory [io.quarkus.test.junit.QuarkusTestExtension] failed to instantiate test class [org.springframework.samples.petclinic.service.userService.UserServiceTests]
Caused by: java.lang.ExceptionInInitializerError
Caused by: java.lang.RuntimeException: Failed to start quarkus
Caused by: java.lang.RuntimeException: Failed to initialize Arc
Caused by: java.lang.ClassFormatError: Duplicate method name "findById" with signature "(Ljava.lang.Object;)Ljava.lang.Object;" in class file org/springframework/samples/petclinic/repository/jpa/JpaOwnerRepository
```

**Expected behavior**
The overrided method should take place, having only one method, not 2.

**Actual behavior**
Apparently 2 methods go to the compiled artifact

**To Reproduce**
Steps to reproduce the behavior:
1. Create a class implementing PanacheRepositoryBase<T, M>
2. Override *findById* method with the same identity type
3. Build and execute tests

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: Linux 192.168.1.134 4.20.15-200.fc29.x86_64 #1 SMP Mon Mar 11 16:01:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

 - Output of `java -version`:  openjdk version "1.8.0_181-ojdkbuild"

 - GraalVM version (if different from Java): 19.2.1
 - Quarkus version or git rev: master branch at today

--- a/extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheJpaEntityEnhancer.java
+++ b/extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheJpaEntityEnhancer.java
@@ -45,14 +45,17 @@ public class PanacheJpaEntityEnhancer extends PanacheEntityEnhancer<MetamodelInf
 
     @Override
     public ClassVisitor apply(String className, ClassVisitor outputClassVisitor) {
-        return new PanacheJpaEntityClassVisitor(className, outputClassVisitor, modelInfo, panacheEntityBaseClassInfo);
+        return new PanacheJpaEntityClassVisitor(className, outputClassVisitor, modelInfo, panacheEntityBaseClassInfo,
+                indexView.getClassByName(DotName.createSimple(className)));
     }
 
     static class PanacheJpaEntityClassVisitor extends PanacheEntityClassVisitor<EntityField> {
 
         public PanacheJpaEntityClassVisitor(String className, ClassVisitor outputClassVisitor,
-                MetamodelInfo<EntityModel<EntityField>> modelInfo, ClassInfo panacheEntityBaseClassInfo) {
-            super(className, outputClassVisitor, modelInfo, panacheEntityBaseClassInfo);
+                MetamodelInfo<EntityModel<EntityField>> modelInfo,
+                ClassInfo panacheEntityBaseClassInfo,
+                ClassInfo entityInfo) {
+            super(className, outputClassVisitor, modelInfo, panacheEntityBaseClassInfo, entityInfo);
         }
 
         @Override

--- a/extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheJpaEntityEnhancer.java
+++ b/extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheJpaEntityEnhancer.java
@@ -25,16 +25,16 @@ import io.quarkus.panache.common.deployment.PanacheEntityEnhancer;
 
 public class PanacheJpaEntityEnhancer extends PanacheEntityEnhancer<MetamodelInfo<EntityModel<EntityField>>> {
 
-    public final static String ENTITY_BASE_NAME = PanacheEntityBase.class.getName();
-    public final static String ENTITY_BASE_BINARY_NAME = ENTITY_BASE_NAME.replace('.', '/');
-    public final static String ENTITY_BASE_SIGNATURE = "L" + ENTITY_BASE_BINARY_NAME + ";";
+    public static final String ENTITY_BASE_NAME = PanacheEntityBase.class.getName();
+    public static final String ENTITY_BASE_BINARY_NAME = ENTITY_BASE_NAME.replace('.', '/');
+    public static final String ENTITY_BASE_SIGNATURE = "L" + ENTITY_BASE_BINARY_NAME + ";";
 
-    public final static String QUERY_NAME = PanacheQuery.class.getName();
-    public final static String QUERY_BINARY_NAME = QUERY_NAME.replace('.', '/');
-    public final static String QUERY_SIGNATURE = "L" + QUERY_BINARY_NAME + ";";
+    public static final String QUERY_NAME = PanacheQuery.class.getName();
+    public static final String QUERY_BINARY_NAME = QUERY_NAME.replace('.', '/');
+    public static final String QUERY_SIGNATURE = "L" + QUERY_BINARY_NAME + ";";
 
-    public final static String JPA_OPERATIONS_NAME = JpaOperations.class.getName();
-    public final static String JPA_OPERATIONS_BINARY_NAME = JPA_OPERATIONS_NAME.replace('.', '/');
+    public static final String JPA_OPERATIONS_NAME = JpaOperations.class.getName();
+    public static final String JPA_OPERATIONS_BINARY_NAME = JPA_OPERATIONS_NAME.replace('.', '/');
 
     private static final DotName DOTNAME_TRANSIENT = DotName.createSimple(Transient.class.getName());
 

--- a/extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheJpaRepositoryEnhancer.java
+++ b/extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheJpaRepositoryEnhancer.java
@@ -9,6 +9,7 @@ import org.objectweb.asm.Opcodes;
 
 import io.quarkus.hibernate.orm.panache.PanacheRepository;
 import io.quarkus.hibernate.orm.panache.PanacheRepositoryBase;
+import io.quarkus.panache.common.deployment.JandexUtil;
 import io.quarkus.panache.common.deployment.PanacheRepositoryEnhancer;
 
 public class PanacheJpaRepositoryEnhancer extends PanacheRepositoryEnhancer {
@@ -54,41 +55,49 @@ public class PanacheJpaRepositoryEnhancer extends PanacheRepositoryEnhancer {
             // Bridge for findById, but only if we actually know the end entity (which we don't for intermediate
             // abstract repositories that haven't fixed their entity type yet
             if (!"Ljava/lang/Object;".equals(entitySignature)) {
-                MethodVisitor mv = super.visitMethod(Opcodes.ACC_PUBLIC | Opcodes.ACC_SYNTHETIC | Opcodes.ACC_BRIDGE,
-                        "findById",
-                        "(Ljava/lang/Object;)Ljava/lang/Object;",
-                        null,
-                        null);
-                mv.visitParameter("id", 0);
-                mv.visitCode();
-                mv.visitIntInsn(Opcodes.ALOAD, 0);
-                mv.visitIntInsn(Opcodes.ALOAD, 1);
-                mv.visitMethodInsn(Opcodes.INVOKEVIRTUAL,
-                        daoBinaryName,
-                        "findById",
-                        "(Ljava/lang/Object;)" + entitySignature, false);
-                mv.visitInsn(Opcodes.ARETURN);
-                mv.visitMaxs(0, 0);
-                mv.visitEnd();
+                if (!JandexUtil.containsMethod(daoClassInfo, "findById",
+                        Object.class.getName(),
+                        Object.class.getName())) {
+                    MethodVisitor mv = super.visitMethod(Opcodes.ACC_PUBLIC | Opcodes.ACC_SYNTHETIC | Opcodes.ACC_BRIDGE,
+                            "findById",
+                            "(Ljava/lang/Object;)Ljava/lang/Object;",
+                            null,
+                            null);
+                    mv.visitParameter("id", 0);
+                    mv.visitCode();
+                    mv.visitIntInsn(Opcodes.ALOAD, 0);
+                    mv.visitIntInsn(Opcodes.ALOAD, 1);
 
-                mv = super.visitMethod(Opcodes.ACC_PUBLIC | Opcodes.ACC_SYNTHETIC | Opcodes.ACC_BRIDGE,
-                        "findById",
-                        "(Ljava/lang/Object;Ljavax/persistence/LockModeType;)Ljava/lang/Object;",
-                        null,
-                        null);
-                mv.visitParameter("id", 0);
-                mv.visitParameter("lockModeType", 0);
-                mv.visitCode();
-                mv.visitIntInsn(Opcodes.ALOAD, 0);
-                mv.visitIntInsn(Opcodes.ALOAD, 1);
-                mv.visitIntInsn(Opcodes.ALOAD, 2);
-                mv.visitMethodInsn(Opcodes.INVOKEVIRTUAL,
-                        daoBinaryName,
-                        "findById",
-                        "(Ljava/lang/Object;Ljavax/persistence/LockModeType;)" + entitySignature, false);
-                mv.visitInsn(Opcodes.ARETURN);
-                mv.visitMaxs(0, 0);
-                mv.visitEnd();
+                    mv.visitMethodInsn(Opcodes.INVOKEVIRTUAL,
+                            daoBinaryName,
+                            "findById",
+                            "(Ljava/lang/Object;)" + entitySignature, false);
+                    mv.visitInsn(Opcodes.ARETURN);
+                    mv.visitMaxs(0, 0);
+                    mv.visitEnd();
+                }
+                if (!JandexUtil.containsMethod(daoClassInfo, "findById",
+                        Object.class.getName(),
+                        Object.class.getName(), "javax.persistence.LockModeType")) {
+                    MethodVisitor mv = super.visitMethod(Opcodes.ACC_PUBLIC | Opcodes.ACC_SYNTHETIC | Opcodes.ACC_BRIDGE,
+                            "findById",
+                            "(Ljava/lang/Object;Ljavax/persistence/LockModeType;)Ljava/lang/Object;",
+                            null,
+                            null);
+                    mv.visitParameter("id", 0);
+                    mv.visitParameter("lockModeType", 0);
+                    mv.visitCode();
+                    mv.visitIntInsn(Opcodes.ALOAD, 0);
+                    mv.visitIntInsn(Opcodes.ALOAD, 1);
+                    mv.visitIntInsn(Opcodes.ALOAD, 2);
+                    mv.visitMethodInsn(Opcodes.INVOKEVIRTUAL,
+                            daoBinaryName,
+                            "findById",
+                            "(Ljava/lang/Object;Ljavax/persistence/LockModeType;)" + entitySignature, false);
+                    mv.visitInsn(Opcodes.ARETURN);
+                    mv.visitMaxs(0, 0);
+                    mv.visitEnd();
+                }
             }
             super.visitEnd();
         }

--- a/extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheJpaRepositoryEnhancer.java
+++ b/extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheJpaRepositoryEnhancer.java
@@ -14,8 +14,8 @@ import io.quarkus.panache.common.deployment.PanacheRepositoryEnhancer;
 
 public class PanacheJpaRepositoryEnhancer extends PanacheRepositoryEnhancer {
 
-    private final static DotName PANACHE_REPOSITORY_BINARY_NAME = DotName.createSimple(PanacheRepository.class.getName());
-    private final static DotName PANACHE_REPOSITORY_BASE_BINARY_NAME = DotName
+    private static final DotName PANACHE_REPOSITORY_BINARY_NAME = DotName.createSimple(PanacheRepository.class.getName());
+    private static final DotName PANACHE_REPOSITORY_BASE_BINARY_NAME = DotName
             .createSimple(PanacheRepositoryBase.class.getName());
 
     public PanacheJpaRepositoryEnhancer(IndexView index) {

--- a/extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheMongoEntityEnhancer.java
+++ b/extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheMongoEntityEnhancer.java
@@ -35,14 +35,16 @@ public class PanacheMongoEntityEnhancer extends PanacheEntityEnhancer<MetamodelI
 
     @Override
     public ClassVisitor apply(String className, ClassVisitor outputClassVisitor) {
-        return new PanacheMongoEntityClassVisitor(className, outputClassVisitor, modelInfo, panacheEntityBaseClassInfo);
+        return new PanacheMongoEntityClassVisitor(className, outputClassVisitor, modelInfo, panacheEntityBaseClassInfo,
+                indexView.getClassByName(DotName.createSimple(className)));
     }
 
     static class PanacheMongoEntityClassVisitor extends PanacheEntityClassVisitor<EntityField> {
 
         public PanacheMongoEntityClassVisitor(String className, ClassVisitor outputClassVisitor,
-                MetamodelInfo<EntityModel<EntityField>> modelInfo, ClassInfo panacheEntityBaseClassInfo) {
-            super(className, outputClassVisitor, modelInfo, panacheEntityBaseClassInfo);
+                MetamodelInfo<EntityModel<EntityField>> modelInfo, ClassInfo panacheEntityBaseClassInfo,
+                ClassInfo entityInfo) {
+            super(className, outputClassVisitor, modelInfo, panacheEntityBaseClassInfo, entityInfo);
         }
 
         @Override

--- a/extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheMongoRepositoryEnhancer.java
+++ b/extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheMongoRepositoryEnhancer.java
@@ -9,6 +9,7 @@ import org.objectweb.asm.Opcodes;
 
 import io.quarkus.mongodb.panache.PanacheMongoRepository;
 import io.quarkus.mongodb.panache.PanacheMongoRepositoryBase;
+import io.quarkus.panache.common.deployment.JandexUtil;
 import io.quarkus.panache.common.deployment.PanacheRepositoryEnhancer;
 
 public class PanacheMongoRepositoryEnhancer extends PanacheRepositoryEnhancer {
@@ -53,22 +54,26 @@ public class PanacheMongoRepositoryEnhancer extends PanacheRepositoryEnhancer {
             // Bridge for findById, but only if we actually know the end entity (which we don't for intermediate
             // abstract repositories that haven't fixed their entity type yet
             if (!entitySignature.equals("Ljava/lang/Object;")) {
-                MethodVisitor mv = super.visitMethod(Opcodes.ACC_PUBLIC | Opcodes.ACC_SYNTHETIC | Opcodes.ACC_BRIDGE,
-                        "findById",
-                        "(Ljava/lang/Object;)Ljava/lang/Object;",
-                        null,
-                        null);
-                mv.visitParameter("id", 0);
-                mv.visitCode();
-                mv.visitIntInsn(Opcodes.ALOAD, 0);
-                mv.visitIntInsn(Opcodes.ALOAD, 1);
-                mv.visitMethodInsn(Opcodes.INVOKEVIRTUAL,
-                        daoBinaryName,
-                        "findById",
-                        "(Ljava/lang/Object;)" + entitySignature, false);
-                mv.visitInsn(Opcodes.ARETURN);
-                mv.visitMaxs(0, 0);
-                mv.visitEnd();
+                if (!JandexUtil.containsMethod(daoClassInfo, "findById",
+                        Object.class.getName(),
+                        Object.class.getName())) {
+                    MethodVisitor mv = super.visitMethod(Opcodes.ACC_PUBLIC | Opcodes.ACC_SYNTHETIC | Opcodes.ACC_BRIDGE,
+                            "findById",
+                            "(Ljava/lang/Object;)Ljava/lang/Object;",
+                            null,
+                            null);
+                    mv.visitParameter("id", 0);
+                    mv.visitCode();
+                    mv.visitIntInsn(Opcodes.ALOAD, 0);
+                    mv.visitIntInsn(Opcodes.ALOAD, 1);
+                    mv.visitMethodInsn(Opcodes.INVOKEVIRTUAL,
+                            daoBinaryName,
+                            "findById",
+                            "(Ljava/lang/Object;)" + entitySignature, false);
+                    mv.visitInsn(Opcodes.ARETURN);
+                    mv.visitMaxs(0, 0);
+                    mv.visitEnd();
+                }
             }
 
             super.visitEnd();

--- a/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/JandexUtil.java
+++ b/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/JandexUtil.java
@@ -187,4 +187,18 @@ public class JandexUtil {
     static Type toClassType(String type) {
         return Type.create(DotName.createSimple(type), Type.Kind.CLASS);
     }
+
+    public static boolean containsMethod(ClassInfo classInfo, MethodInfo methodInfo) {
+        if (classInfo.methods().contains(methodInfo)) {
+            return true;
+        }
+        // MethodInfo may not belong to the same declaring class. Check signature
+        for (MethodInfo classMethodInfo : classInfo.methods()) {
+            if (classMethodInfo.name().equals(methodInfo.name()) &&
+                    classMethodInfo.parameters().equals(methodInfo.parameters())) {
+                return true;
+            }
+        }
+        return false;
+    }
 }

--- a/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/JandexUtil.java
+++ b/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/JandexUtil.java
@@ -1,9 +1,13 @@
 package io.quarkus.panache.common.deployment;
 
+import static java.util.stream.Collectors.toList;
+
+import java.util.Arrays;
 import java.util.List;
 import java.util.function.Function;
 
 import org.jboss.jandex.ArrayType;
+import org.jboss.jandex.ClassInfo;
 import org.jboss.jandex.DotName;
 import org.jboss.jandex.MethodInfo;
 import org.jboss.jandex.ParameterizedType;
@@ -159,4 +163,28 @@ public class JandexUtil {
         }
     }
 
+    /**
+     * Checks if the {@link ClassInfo} contains a method
+     *
+     * @param classInfo the {@link ClassInfo} instance
+     * @param methodName the method name to check
+     * @param parameters the parameter types, if any
+     * @return true if the {@link ClassInfo} parameter contains this method
+     */
+    public static boolean containsMethod(ClassInfo classInfo,
+            String methodName,
+            String returnType,
+            String... parameters) {
+        List<Type> types = Arrays.stream(parameters).map(JandexUtil::toClassType).collect(toList());
+        for (MethodInfo methodInfo : classInfo.methods()) {
+            if (methodInfo.name().equals(methodName) && methodInfo.parameters().equals(types)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    static Type toClassType(String type) {
+        return Type.create(DotName.createSimple(type), Type.Kind.CLASS);
+    }
 }

--- a/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheEntityEnhancer.java
+++ b/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheEntityEnhancer.java
@@ -40,9 +40,11 @@ public abstract class PanacheEntityEnhancer<MetamodelType extends MetamodelInfo<
 
     protected MetamodelType modelInfo;
     protected final ClassInfo panacheEntityBaseClassInfo;
+    protected final IndexView indexView;
 
     public PanacheEntityEnhancer(IndexView index, DotName panacheEntityBaseName) {
-        panacheEntityBaseClassInfo = index.getClassByName(panacheEntityBaseName);
+        this.panacheEntityBaseClassInfo = index.getClassByName(panacheEntityBaseName);
+        this.indexView = index;
     }
 
     @Override
@@ -56,24 +58,28 @@ public abstract class PanacheEntityEnhancer<MetamodelType extends MetamodelInfo<
         private Set<String> methods = new HashSet<>();
         private MetamodelInfo<?> modelInfo;
         private ClassInfo panacheEntityBaseClassInfo;
+        protected ClassInfo entityInfo;
 
         public PanacheEntityClassVisitor(String className, ClassVisitor outputClassVisitor,
                 MetamodelInfo<? extends EntityModel<? extends EntityFieldType>> modelInfo,
-                ClassInfo panacheEntityBaseClassInfo) {
+                ClassInfo panacheEntityBaseClassInfo,
+                ClassInfo entityInfo) {
             super(Opcodes.ASM7, outputClassVisitor);
             thisClass = Type.getType("L" + className.replace('.', '/') + ";");
             this.modelInfo = modelInfo;
             EntityModel<? extends EntityFieldType> entityModel = modelInfo.getEntityModel(className);
             fields = entityModel != null ? entityModel.fields : null;
             this.panacheEntityBaseClassInfo = panacheEntityBaseClassInfo;
+            this.entityInfo = entityInfo;
         }
 
         @Override
         public FieldVisitor visitField(int access, String name, String descriptor, String signature, Object value) {
             FieldVisitor superVisitor = super.visitField(access, name, descriptor, signature, value);
             EntityField ef = fields.get(name);
-            if (fields == null || ef == null)
+            if (ef == null) {
                 return superVisitor;
+            }
             ef.signature = signature;
             // if we have a mapped field, let's add some annotations
             return new FieldVisitor(Opcodes.ASM7, superVisitor) {
@@ -107,8 +113,9 @@ public abstract class PanacheEntityEnhancer<MetamodelType extends MetamodelInfo<
                 String[] exceptions) {
             if (methodName.startsWith("get")
                     || methodName.startsWith("set")
-                    || methodName.startsWith("is"))
+                    || methodName.startsWith("is")) {
                 methods.add(methodName + "/" + descriptor);
+            }
             MethodVisitor superVisitor = super.visitMethod(access, methodName, descriptor, signature, exceptions);
             return new PanacheFieldAccessMethodVisitor(superVisitor, thisClass.getInternalName(), methodName, descriptor,
                     modelInfo);
@@ -119,9 +126,13 @@ public abstract class PanacheEntityEnhancer<MetamodelType extends MetamodelInfo<
             // FIXME: generate default constructor
 
             for (MethodInfo method : panacheEntityBaseClassInfo.methods()) {
-                AnnotationInstance bridge = method.annotation(JandexUtil.DOTNAME_GENERATE_BRIDGE);
-                if (bridge != null)
-                    generateMethod(method, bridge.value("targetReturnTypeErased"));
+                // Do not generate a method that already exists
+                if (!JandexUtil.containsMethod(entityInfo, method)) {
+                    AnnotationInstance bridge = method.annotation(JandexUtil.DOTNAME_GENERATE_BRIDGE);
+                    if (bridge != null) {
+                        generateMethod(method, bridge.value("targetReturnTypeErased"));
+                    }
+                }
             }
 
             generateAccessors();

--- a/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheRepositoryEnhancer.java
+++ b/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheRepositoryEnhancer.java
@@ -106,7 +106,7 @@ public abstract class PanacheRepositoryEnhancer implements BiFunction<String, Cl
         public void visitEnd() {
             for (MethodInfo method : panacheRepositoryBaseClassInfo.methods()) {
                 // Do not generate a method that already exists
-                if (!daoClassInfo.methods().contains(method)) {
+                if (!JandexUtil.containsMethod(daoClassInfo, method)) {
                     AnnotationInstance bridge = method.annotation(JandexUtil.DOTNAME_GENERATE_BRIDGE);
                     if (bridge != null) {
                         generateMethod(method, bridge.value("targetReturnTypeErased"));

--- a/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheRepositoryEnhancer.java
+++ b/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheRepositoryEnhancer.java
@@ -36,12 +36,14 @@ public abstract class PanacheRepositoryEnhancer implements BiFunction<String, Cl
         protected String entitySignature;
         protected String entityBinaryType;
         protected String daoBinaryName;
+        protected ClassInfo daoClassInfo;
         protected ClassInfo panacheRepositoryBaseClassInfo;
         protected IndexView indexView;
 
         public PanacheRepositoryClassVisitor(String className, ClassVisitor outputClassVisitor,
                 ClassInfo panacheRepositoryBaseClassInfo, IndexView indexView) {
             super(Opcodes.ASM7, outputClassVisitor);
+            daoClassInfo = indexView.getClassByName(DotName.createSimple(className));
             daoBinaryName = className.replace('.', '/');
             this.panacheRepositoryBaseClassInfo = panacheRepositoryBaseClassInfo;
             this.indexView = indexView;
@@ -101,20 +103,16 @@ public abstract class PanacheRepositoryEnhancer implements BiFunction<String, Cl
         }
 
         @Override
-        public MethodVisitor visitMethod(int access, String name, String descriptor, String signature, String[] exceptions) {
-            // FIXME: do not add method if already present
-            return super.visitMethod(access, name, descriptor, signature, exceptions);
-        }
-
-        @Override
         public void visitEnd() {
-
             for (MethodInfo method : panacheRepositoryBaseClassInfo.methods()) {
-                AnnotationInstance bridge = method.annotation(JandexUtil.DOTNAME_GENERATE_BRIDGE);
-                if (bridge != null)
-                    generateMethod(method, bridge.value("targetReturnTypeErased"));
+                // Do not generate a method that already exists
+                if (!daoClassInfo.methods().contains(method)) {
+                    AnnotationInstance bridge = method.annotation(JandexUtil.DOTNAME_GENERATE_BRIDGE);
+                    if (bridge != null) {
+                        generateMethod(method, bridge.value("targetReturnTypeErased"));
+                    }
+                }
             }
-
             super.visitEnd();
         }
 
@@ -126,9 +124,9 @@ public abstract class PanacheRepositoryEnhancer implements BiFunction<String, Cl
             String castTo = null;
             if (targetReturnTypeErased != null && targetReturnTypeErased.asBoolean()) {
                 org.jboss.jandex.Type type = method.returnType();
-                if (type.kind() == Kind.TYPE_VARIABLE) {
-                    if (type.asTypeVariable().identifier().equals("Entity"))
-                        castTo = entityBinaryType;
+                if (type.kind() == Kind.TYPE_VARIABLE &&
+                        type.asTypeVariable().identifier().equals("Entity")) {
+                    castTo = entityBinaryType;
                 }
                 if (castTo == null)
                     castTo = type.name().toString('/');

--- a/integration-tests/hibernate-orm-panache/pom.xml
+++ b/integration-tests/hibernate-orm-panache/pom.xml
@@ -68,6 +68,11 @@
             <artifactId>quarkus-jackson</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <scope>test</scope>
+        </dependency>
     </dependencies>
 
     <build>

--- a/integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/DuplicateEntity.java
+++ b/integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/DuplicateEntity.java
@@ -1,0 +1,26 @@
+package io.quarkus.it.panache;
+
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.Id;
+
+import io.quarkus.hibernate.orm.panache.PanacheEntityBase;
+
+@Entity
+public class DuplicateEntity extends PanacheEntityBase {
+
+    @Id
+    @GeneratedValue
+    public Integer id;
+
+    public static DuplicateEntity findById(Object id) {
+        DuplicateEntity duplicate = new DuplicateEntity();
+        duplicate.id = (Integer) id;
+        return duplicate;
+    }
+
+    @Override
+    public void persist() {
+        // Do nothing
+    }
+}

--- a/integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/DuplicateEntity.java
+++ b/integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/DuplicateEntity.java
@@ -5,6 +5,7 @@ import javax.persistence.GeneratedValue;
 import javax.persistence.Id;
 
 import io.quarkus.hibernate.orm.panache.PanacheEntityBase;
+import io.quarkus.panache.common.Parameters;
 
 @Entity
 public class DuplicateEntity extends PanacheEntityBase {
@@ -13,14 +14,18 @@ public class DuplicateEntity extends PanacheEntityBase {
     @GeneratedValue
     public Integer id;
 
-    public static DuplicateEntity findById(Object id) {
+    public static <T extends PanacheEntityBase> T findById(Object id) {
         DuplicateEntity duplicate = new DuplicateEntity();
         duplicate.id = (Integer) id;
-        return duplicate;
+        return (T) duplicate;
     }
 
     @Override
     public void persist() {
         // Do nothing
+    }
+
+    public static int update(String query, Parameters params) {
+        return 0;
     }
 }

--- a/integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/DuplicateRepository.java
+++ b/integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/DuplicateRepository.java
@@ -1,0 +1,16 @@
+package io.quarkus.it.panache;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import io.quarkus.hibernate.orm.panache.PanacheRepositoryBase;
+
+@ApplicationScoped
+public class DuplicateRepository implements PanacheRepositoryBase<DuplicateEntity, Integer> {
+
+    @Override
+    public DuplicateEntity findById(Integer id) {
+        DuplicateEntity duplicate = new DuplicateEntity();
+        duplicate.id = id;
+        return duplicate;
+    }
+}

--- a/integration-tests/hibernate-orm-panache/src/test/java/io/quarkus/it/panache/DuplicateMethodTest.java
+++ b/integration-tests/hibernate-orm-panache/src/test/java/io/quarkus/it/panache/DuplicateMethodTest.java
@@ -1,0 +1,29 @@
+package io.quarkus.it.panache;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatCode;
+
+import javax.inject.Inject;
+
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.test.junit.QuarkusTest;
+
+@QuarkusTest
+public class DuplicateMethodTest {
+
+    @Inject
+    DuplicateRepository repository;
+
+    @Test
+    public void shouldNotDuplicateMethodsInRepository() {
+        assertThat(repository.findById(1)).isNotNull();
+    }
+
+    @Test
+    public void shouldNotDuplicateMethodsInEntity() {
+        DuplicateEntity entity = DuplicateEntity.findById(1);
+        assertThat(entity).isNotNull();
+        assertThatCode(entity::persist).doesNotThrowAnyException();
+    }
+}

--- a/integration-tests/hibernate-orm-panache/src/test/java/io/quarkus/it/panache/DuplicateMethodTest.java
+++ b/integration-tests/hibernate-orm-panache/src/test/java/io/quarkus/it/panache/DuplicateMethodTest.java
@@ -7,6 +7,7 @@ import javax.inject.Inject;
 
 import org.junit.jupiter.api.Test;
 
+import io.quarkus.panache.common.Parameters;
 import io.quarkus.test.junit.QuarkusTest;
 
 @QuarkusTest
@@ -25,5 +26,7 @@ public class DuplicateMethodTest {
         DuplicateEntity entity = DuplicateEntity.findById(1);
         assertThat(entity).isNotNull();
         assertThatCode(entity::persist).doesNotThrowAnyException();
+        assertThatCode(() -> DuplicateEntity.update("foo", Parameters.with("a", 1)))
+                .doesNotThrowAnyException();
     }
 }

